lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,311 @@
+            , sharedStateAddressPool =
+                mkAddressPool @n (ParentContextMultisigScript accXPub pT dT) g []
+            }
+        else
+            st
+
+templatesComplete
+    :: ScriptTemplate
+    -> Maybe ScriptTemplate
+    -> Bool
+templatesComplete pTemplate dTemplate =
+    isRight (validateScriptTemplate RequiredValidation pTemplate) &&
+    case dTemplate of
+        Nothing -> True
+        Just dTemplate' ->
+            isRight (validateScriptTemplate RequiredValidation dTemplate')
+
+-- | The cosigner with his account public key is done per template.
+-- For every template the script is checked if the cosigner is present.
+-- If yes, then the key is inserted. If the key is already present it is going to be
+-- updated. If there is no cosigner present is the script then the cosigner -
+-- account public key map is not changed. The updating works with pending shared state,
+-- and can unpend the shared state if all public keys are present. When already unpended
+-- shared state is used it does not change it.
+addCosignerAccXPub
+    :: forall (n :: NetworkDiscriminant) k.
+       ( MkKeyFingerprint k Address
+       , MkKeyFingerprint k (Proxy n, k 'AddressK XPub)
+       , SoftDerivation k
+       , Typeable n
+       , WalletKey k )
+    => k 'AccountK XPub
+    -> Cosigner
+    -> SharedState n k
+    -> SharedState n k
+addCosignerAccXPub accXPub cosigner state = case state of
+    st@(SharedState _ _) -> st
+    PendingSharedState prefix accXPub' pT dT g ->
+        let updateScriptTemplate sc@(ScriptTemplate cosignerMap script') =
+                if cosigner `elem` retrieveAllCosigners script' then
+                    ScriptTemplate (Map.insert cosigner (getRawKey accXPub) cosignerMap) script'
+                else
+                    sc
+            st = PendingSharedState
+                { pendingSharedStateDerivationPrefix = prefix
+                , pendingSharedStateAccountKey = accXPub'
+                , pendingSharedStatePaymentTemplate = updateScriptTemplate pT
+                , pendingSharedStateDelegationTemplate = updateScriptTemplate <$> dT
+                , pendingSharedStateAddressPoolGap = g
+                }
+        in trySharedState st
+
+retrieveAllCosigners :: Script Cosigner -> [Cosigner]
+retrieveAllCosigners = foldScript (:) []
+
+isShared
+    :: forall (n :: NetworkDiscriminant) k.
+       ( SoftDerivation k
+       , WalletKey k
+       , Typeable n
+       , MkKeyFingerprint k Address
+       , MkKeyFingerprint k (Proxy n, k 'AddressK XPub) )
+    => Address
+    -> SharedState n k
+    -> (Maybe (Index 'Soft 'ScriptK, KeyHash), SharedState n k)
+isShared addr st = case st of
+    SharedState prefix pool ->
+        let (ixM, pool') = lookupAddress @n (const Used) addr pool
+            (ParentContextMultisigScript accXPub _ _) = context pool
+        in case ixM of
+            Just ix ->
+                (Just (coerce ix, keyHashFromAccXPubIx accXPub (coerce ix))
```

Comment:
What does this hash corresponds to? (I mean, where do we intend to use it?)

Reply:
this is the verification key corresponding to `ix` and `accXPub` that was found in discovered script address we co-share